gnovm/cmd/gno/transpile.go

```diff
@@ -259,8 +292,116 @@ func goBuildFileOrPkg(fileOrPkg string, cfg *transpileCfg) error {
+
+			dirs = append(dirs, rootDir+filepath.FromSlash(res))
+		}
+	}
+	return
+}
+
+// buildTranspiledPackage tries to run `go build` against the transpiled .go files.
+//
+// This method is the most efficient to detect errors but requires that
+// all the import are valid and available.
+func buildTranspiledPackage(fileOrPkg, goBinary string) error {
+	// TODO: use cmd/compile instead of exec?
+	// TODO: find the nearest go.mod file, chdir in the same folder, trim prefix?
+	// TODO: temporarily create an in-memory go.mod or disable go modules for gno?
+	// TODO: ignore .go files that were not generated from gno?
+
+	info, err := os.Stat(fileOrPkg)
+	if err != nil {
+		return fmt.Errorf("invalid file or package path %s: %w", fileOrPkg, err)
+	}
+	var (
+		target string
+		chdir  string
+	)
+	if !info.IsDir() {
+		dstFilename, _ := transpiler.TranspiledFilenameAndTags(fileOrPkg)
+		// Makes clear to go compiler that this is a relative path,
+		// rather than a path to a package/module.
+		// can't use filepath.Join as it cleans its results.
+		target = filepath.Dir(fileOrPkg) + string(filepath.Separator) + dstFilename
+	} else {
+		// Go does not allow building packages using absolute paths, and requires
+		// relative paths to always be prefixed with `./` (because the argument
+		// go expects are import paths, not directories).
+		// To circumvent this, we use the -C flag to chdir into the right
+		// directory, then run `go build .`
+		chdir = fileOrPkg
+		target = "."
+	}
+
+	// pre-alloc max 5 args
+	args := append(make([]string, 0, 5), "build")
+	if chdir != "" {
+		args = append(args, "-C", chdir)
+	}
+	args = append(args, "-tags=gno", target)
+	cmd := exec.Command(goBinary, args...)
+	out, err := cmd.CombinedOutput()
+	if _, ok := err.(*exec.ExitError); ok {
+		// exit error
+		return parseGoBuildErrors(string(out))
+	}
+	return err
+}
+
+var (
+	reGoBuildError   = regexp.MustCompile(`(?m)^(\S+):(\d+):(\d+): (.+)$`)
+	reGoBuildComment = regexp.MustCompile(`(?m)^#.*$`)
+)
+
+// parseGoBuildErrors returns a scanner.ErrorList filled with all errors found
+// in out, which is supposed to be the output of the `go build` command.
+//
+// TODO(tb): update when `go build -json` is released to replace regexp usage.
+// See https://github.com/golang/go/issues/62067
+func parseGoBuildErrors(out string) error {
+	var errList scanner.ErrorList
+	matches := reGoBuildError.FindAllStringSubmatch(out, -1)
+	for _, match := range matches {
+		filename := match[1]
+		line, err := strconv.Atoi(match[2])
+		if err != nil {
+			return fmt.Errorf("parse line go build error %s: %w", match, err)
+		}
+
+		column, err := strconv.Atoi(match[3])
+		if err != nil {
+			return fmt.Errorf("parse column go build error %s: %w", match, err)
+		}
+		msg := match[4]
+		errList.Add(token.Position{
+			Filename: filename,
+			Line:     line,
+			Column:   column,
+		}, msg)
+	}
+
+	replaced := reGoBuildError.ReplaceAllLiteralString(out, "")
```

Comment:
We don't cover these lines in the unit tests for this func

Reply:
I blame t.Parallel

Fixed